deployments: add some basic tests for npm package

compojoom opened the pull request:
It happened that we had an empty abis/mainnet folder - which caused a broken npm release. The test here would prevent something like this in the future.

@istankovic - running the test out of the make `npm-package` would stop the release of pytest errors out, right?

compojoom commented:
I also think we should maybe change the circleCI conf:

https://github.com/beamer-bridge/beamer/blob/deployments/add_tests_for_deployment/.circleci/workflows.yml#L324

Instead of creating the package only when we have modified artifacts or abi, we maybe should run the `make npm-package` all the time - this way it would detect if something in the pipeline(generate-abi.py for example) is broken?

istankovic commented on deployments/tests/test_dist.py:
```diff
@@ -0,0 +1,34 @@
+    assert len(json_files) > 0, f'No JSON files in {directory}'
+
+    for file in json_files:
+        assert is_valid_json(os.path.join(directory, file)), f'File {file} in {directory} is not valid JSON'
```
I'm thinking... is there a way to test ("import") the newly generated npm package before publishing?
Because then we could actually have a proper test (JS test) that imports the package and does some basic sanity checks, e.g. for existence of an address of a certain contract etc.

compojoom replied:
hm, good idea. Do I test for example just the RequestManager? Or do we test all currently available contracts?

istankovic replied:
> hm, good idea. Do I test for example just the RequestManager? Or do we test all currently available contracts?

It depends on how deeply you want to test. Personally, I would at least make sure that all contracts have addresses and that there are at least `RequestManager` and `FillManager` there -- that would be sufficient as a smoke test, IMO.

istankovic commented:
> I also think we should maybe change the circleCI conf:
> 
> https://github.com/beamer-bridge/beamer/blob/deployments/add_tests_for_deployment/.circleci/workflows.yml#L324
> 
> Instead of creating the package only when we have modified artifacts or abi, we maybe should run the `make npm-package` all the time - this way it would detect if something in the pipeline(generate-abi.py for example) is broken?

That is a very good point, yes. :+1: 
Whenever the backend is changed, we should run the "build and test" job and if the artifacts are changed, we additionally should also run the "publish" part.

compojoom commented:
I've moved the code for the npm package to a separate directory and added a step in the workflow to execute the test. 

The test generates a package, installs it and we compare whether the expected abis are there (we determine that based on the contract names in the artifacts) and we also check whether all artifacts have been copied to the artifacts folder of the package.

istankovic commented:
Could you please remove commit 5ab9ccf86914fdd2d6c6c7572f58049b385ec0ba from this PR? It's not necessary anymore.

istankovic commented:
Commit f108275cbf83dce89aec0635cde183e19bae1eef has message
```
deployments: add some basic tests for npm package
```
which is wrong, because that commit is about updating the CircleCI workflow.

istankovic commented:
@manuelwedler @GabrielBuragev could you also please take a look at this PR?
I am afraid I don't know enough about JS things.

istankovic commented:
Nitpick: the commit message of the first commit:
```
deployments: move npm packaage to separate folder

We’ve moved the code that generates the package to a separate folder and added some basic tests for whether the package generated contains the expected files.

deployments: wrong files copied
```
mentions "wrong files copied", not sure what that is about, probably a remnant of a rebase, but would be nice to remove it.

Also note the typo in `packaage` in the first line.

Aside from that, looks pretty good! (Note that I left the JS part to @GabrielBuragev and/or @manuelwedler).

compojoom commented:
Isn't that the way to write it? :D
![image](https://github.com/beamer-bridge/beamer/assets/693770/038b8f4b-62d5-4c90-81c0-19686f2b2fc5)

GabrielBuragev commented on deployments-npm-package/__tests__/test_package.ts:
```diff
@@ -0,0 +1,113 @@
+const getContracts = async () => {
+  const contractsPerNetwork: { [key: string]: string[] } = {};
+  try {
+    const artifactItems = await fs.readdir(pathToArtifacts);
```
can we rename `artifactItems` to `networks`?

GabrielBuragev commented on deployments-npm-package/__tests__/test_package.ts:
```diff
@@ -0,0 +1,113 @@
+    const artifactItems = await fs.readdir(pathToArtifacts);
+
+    for (const item of artifactItems) {
+      const artifactPath = path.join(pathToArtifacts, item);
```
If you apply the above change, you would have to refactor the other places where `artifactsItems` occurs. 
We could have something like:
`for (const network of networks) { ... }`

Once you do these changes, you will see things will start making more sense in this code block :D

GabrielBuragev commented on deployments-npm-package/__tests__/test_package.ts:
```diff
@@ -0,0 +1,113 @@
+              contractsPerNetwork[item] = contracts;
+            } else {
+              contractsPerNetwork[item] = [...contractsPerNetwork[item], ...contracts];
+            }
```
We could move the Set initialization here. I believe this will work better and be more understandable:
```js
const contracts = new Set([...baseContracts, ...chainContracts]);
contracts.delete('chain_id');

if (!contractsPerNetwork[item]) {
  contractsPerNetwork[item] = [...contracts];
} else {
  contractsPerNetwork[item] = [...contractsPerNetwork[item], ...contracts];
}
```

By having this, you can remove lines 51-53 below.

GabrielBuragev commented on deployments-npm-package/__tests__/test_package.ts:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @returns {Promise<{ [key: string]: string[] }>}
+ */
+const getContracts = async () => {
```
`getContractsPerNetwork` as a method name here is more applicable i think.

GabrielBuragev commented on deployments-npm-package/__tests__/test_package.ts:
```diff
@@ -0,0 +1,113 @@
+  test('test that artifacts are copied to the package', async () => {
+    const artifactItems = await fs.readdir(pathToArtifacts);
+
+    for (const item of artifactItems) {
```
`const network of networks` here.

GabrielBuragev reviewed (changes requested):
I would like to see the naming inside the tests being addressed/improved. 
It was hard to understand the tests due to having not so easy to understand variable names.
I believe we can make it better by renaming/refactoring some parts.

The rest looks fine!
